File: week8&9/Ariling/prgs-충돌위험 찾기/main.js
function solution(points, routes) {
  var answer = 0;
  let point_map = {};
  // 포인트 위치 담기
  for (let i = 0; i < points.length; i++) {
    const [r, c] = points[i];
    point_map[i + 1] = [r, c];
  }
  let route_arr = [];
  let max_len = -1;
  for (let i = 0; i < routes.length; i++) {
    let [curR, curC] = point_map[routes[i][0]];
    const pos_arr = [[curR, curC]];
    for (let j = 1; j < routes[i].length; j++) {
      // 현재 위치에서 가야할 지점 위치 차이 구하기
      let r_diff = curR - point_map[routes[i][j]][0];
      let c_diff = curC - point_map[routes[i][j]][1];
      // 행 먼저 이동한다고 하니 r_diff부터 처리
      while (r_diff != 0) {
        curR += r_diff < 0 ? 1 : -1;
        r_diff += r_diff > 0 ? -1 : 1;
        pos_arr.push([curR, curC]);
      }
      while (c_diff != 0) {
        curC += c_diff < 0 ? 1 : -1;
        c_diff += c_diff > 0 ? -1 : 1;
        pos_arr.push([curR, curC]);
      }
    }
    if (max_len < pos_arr.length) {
      max_len = pos_arr.length;
    }
    route_arr.push(pos_arr);
  }
  for (let i = 0; i < max_len; i++) {
    // filter와 map을 통해서 이동 시점에서 이동하는 좌표들만 담기
    let values = route_arr.filter((el) => i < el.length).map((el) => el[i]);
    let seen = new Set();
    let duplicate = new Set();
    // set을 활용하여 중복 유무를 따지기
    for (let j = 0; j < values.length; j++) {
      const [r, c] = values[j];
      if (seen.has(`${r}, ${c}`)) {
        duplicate.add(`${r}, ${c}`);
      } else {
        seen.add(`${r}, ${c}`);
      }
    }
    answer += duplicate.size;
  }
  return answer;
}

File: week8&9/Ariling/prgs-충돌위험 찾기/main.py
def solution(points, routes):
    # n개의 포인트가 있고 각각 다른 번호이다.
    # m개의 포인트로 구성되고 할당된 포인트를 순서대로 방문한다.
    # x대이고 1초마다 4방향중 하나로 이동이 가능하다.
    # 항상 최단 경로로 이동하며 r을 우선으로 한다. 즉 상하를 먼저 for문으로 두기
    # 벗어나는 경로는 고려하지 않는다. 같은 좌표에 로봇이 2대 이상 있으면 충돌있다고 한다.
    point_arr = {}
    for i in range(len(points)):
        [r, c] = points[i]
        point_arr[i+1] = [r, c]
    # 짧은 길 담는 곳 , 그 정거장 하면서 하려면 일단 가는 방향을 담는게 맞음..
    route_arr= []
    max_len = -1
    # 일단 위치를 박아놓기...
    for route in routes:
        [cur_r, cur_c] = point_arr[route[0]]
        pos_arr = [(cur_r, cur_c)]
        for i in range(1, len(route)):
            r_diff = cur_r - point_arr[route[i]][0]
            c_diff = cur_c - point_arr[route[i]][1]
            while r_diff != 0:
                cur_r += 1 if r_diff < 0 else -1
                r_diff += -1 if r_diff > 0 else 1
                pos_arr.append((cur_r, cur_c))
            while c_diff != 0:
                cur_c += 1 if c_diff < 0 else -1
                c_diff += -1 if c_diff > 0 else 1
                pos_arr.append((cur_r, cur_c))
        if len(pos_arr) > max_len:
            max_len = len(pos_arr)
        route_arr.append(pos_arr)
    answer = 0
    for i in range(max_len):
        # 이 방법은 힌트를 본 건 아니고 어떻게 중복을 찾을지 물어봐서 활용하였다.
        values = [arr[i] for arr in route_arr if i < len(arr)]
        seen = set()
        duplicates = set()
        for val in values:
            if val in seen:
                duplicates.add(val)
            else:
                seen.add(val)
        answer += len(duplicates)
    return answer